packages/compass-connections/src/stores/connections-store-redux.ts

```diff
@@ -1698,6 +1773,58 @@ const connectWithOptions = (
           }
         );
 
+        // Listen for MongoDB driver events and interpret them as connection steps
+        let topologyDiscoveryCompleted = false;
+        let authenticationCompleted = false;
+
+        dataService.on('topologyDescriptionChanged', () => {
+          if (!topologyDiscoveryCompleted) {
+            topologyDiscoveryCompleted = true;
+            dispatch(
+              updateConnectionStep(connectionInfo.id, 'topology', 'completed')
+            );
+            dispatch(
+              updateConnectionStep(
+                connectionInfo.id,
+                'authentication',
+                'in-progress'
+              )
+            );
```

Comment:
Dispatching actions (especially setter actions) in bulk is an [antipattern](https://redux.js.org/style-guide#model-actions-as-events-not-setters) that usually points to a wrongly selected abstraction for the action: actions should roughly map to actual events happening in the app, and then it's reducer job to define what exactly that means. As you are actually dealing with already existing events here, I think the best thing you can do is to actually model your actions around those and make reducer decide how to modify state based on that